elastic/kibana
# [Runtime field editor] Add server side for preview functionality

This PR contains the core logic of the preview functionality of runtime fields inside the index pattern field editor.
It does not contain any UI yet so I output for now the result **in the browser console**.

## How to test

* Add Kibana sample data logs
* Navigate to the index pattern management and select the sample data logs index pattern
* Click on "Add field" and open the Chrome dev tools to see the console
* Give a name to the field (e.g "foo")
* Enable the "Set value" toggle and add a script (e.g. `emit("bar")`)
* You should see the field preview in the console

```
Field preview: {
    "key": "foo",
    "value": "bar"
}
```

* Now update the script with the following where we read data from the index pattern document

```
Long bytesToGB = doc['machine.ram'].value / 1073741824;

emit(bytesToGB + 'GB');
```

* You should see the updated preview

```
Field preview: {
    "key": "foo",
    "value": "8GB"
}
```

## Notes on implementations

### Added `<FieldEditorProvider />`

To avoid property drilling I've added a `fieldEditorContext` to pass down services and context value to components down the tree. In this PR I haven't modified the existing code to reduce the amount of changes. In a [following PR](https://github.com/elastic/kibana/pull/100646) I will refactor to code to make use of the new context.

### New top level `__jest__` folder

In order to follow the pattern put in place in all of our apps I have created a top level `__jest__` folder to add the component integration tests. Now that we have server side endpoint to test, the helpers will help mock the HTTP request response easily and test different scenarios in the UI.
In a [following PR](https://github.com/elastic/kibana/pull/100646) I will refactor the current component integration tests and move them to this folder.

<img width="693" alt="Screenshot 2021-05-17 at 12 27 28" src="https://user-images.githubusercontent.com/2854616/118498483-f78bc900-b71d-11eb-9158-0c8007f7cd49.png">
